File: python/2021/13_transparent_origami/aoc202113.py
"""AoC 13, 2021: Transparent Origami"""

# Standard library imports
import pathlib
import sys

# Third party imports
import advent_of_code_ocr
import colorama
import parse
from colorama import Cursor

colorama.init()

FOLD_PATTERN = parse.compile("fold along {xy}={line:d}")


def parse(puzzle_input):
    """Parse input"""
    dots, _, folds = puzzle_input.partition("\n\n")
    return (parse_dots(dots), parse_folds(folds))


def parse_dots(dots):
    """Parse list of dots"""
    return {(int(x), int(y)) for x, y in [dot.split(",") for dot in dots.split("\n")]}


def parse_folds(folds):
    """Parse list of folds"""
    return [
        (match["xy"], match["line"])
        for match in [FOLD_PATTERN.parse(line) for line in folds.split("\n")]
    ]


def part1(data):
    """Solve part 1"""
    dots, folds = data
    return len(fold(dots, folds[:1]))


def part2(data):
    """Solve part 2"""
    text = draw_dots(fold(*data))

    if "-v" in sys.argv:
        print(text)

    return advent_of_code_ocr.convert_6(text)


def fold(dots, folds):
    """Perform folds on dots

    >>> dots = fold({(0, 0), (1, 1), (3, 0), (4, 0)}, [("x", 2)])
    >>> sorted(dots)
    [(0, 0), (1, 0), (1, 1)]
    """
    for fold in folds:
        match fold:
            case ("x", line):
                dots = {(x, y) if x < line else (2 * line - x, y) for x, y in dots}
            case ("y", line):
                dots = {(x, y) if y < line else (x, 2 * line - y) for x, y in dots}
    return dots


def draw_dots(dots):
    """Create a text string visualizing the dots

    >>> print(draw_dots({(0, 0), (0, 2), (1,0), (1, 1), (1, 2), (2, 0), (2, 2)}))
    ###
    .#.
    ###
    """
    num_rows, num_cols = max(y for _, y in dots), max(x for x, _ in dots)
    rows = []
    for row in range(num_rows + 1):
        cols = {x for x, y in dots if y == row}
        rows.append("".join(".#"[x in cols] for x in range(num_cols + 1)))
    return "\n".join(rows)


def solve(puzzle_input):
    """Solve the puzzle for the given input"""
    data = parse(puzzle_input)
    solution1 = part1(data)
    solution2 = part2(data)

    return solution1, solution2


if __name__ == "__main__":
    for path in sys.argv[1:]:
        if path.startswith("-"):
            continue

        print(f"\n{path}:")
        solutions = solve(puzzle_input=pathlib.Path(path).read_text().strip())
        print("\n".join(str(solution) for solution in solutions))

File: elixir/lib/2021/13_transparent_origami/aoc202113.ex
defmodule AOC2021.Day13 do
  @moduledoc """
  Advent of Code 2021, day 13: Transparent Origami
  """
  require AOC

  @doc """
  Parse input
  """
  def parse(puzzle_input) do
    {dots, folds} =
      puzzle_input
      |> String.split("\n", trim: true)
      |> Enum.split_with(&(not String.starts_with?(&1, "fold along ")))

    {parse_points(dots), parse_folds(folds)}
  end

  @doc """
  Parse a list of dots and store them in a MapSet.

  ## Example:

      iex> parse_points(["1,2", "2,0", "1,2", "99,1"])
      MapSet.new([{1, 2}, {2, 0}, {99, 1}])
  """
  def parse_points(dots) do
    dots
    |> Enum.map(fn line ->
      line |> String.split(",") |> Enum.map(&String.to_integer/1) |> List.to_tuple()
    end)
    |> Enum.into(MapSet.new())
  end

  @doc """
  Parse a list of folds and store them in a keyword list.

  ## Example:

      iex> parse_folds(["fold along y=2", "fold along x=8", "fold along y=1"])
      [y: 2, x: 8, y: 1]
  """
  def parse_folds(folds) do
    folds
    |> Enum.map(fn "fold along " <> <<direction::utf8, "=", line::binary>> ->
      {List.to_atom([direction]), String.to_integer(line)}
    end)
  end

  @doc """
  Solve part 1
  """
  # |> Enum.to_list() |> length()
  def part1({dots, folds}), do: folds |> hd() |> fold_one(dots) |> Enum.to_list() |> length()

  @doc """
  Solve part 2
  """
  def part2({dots, folds}), do: folds |> fold_all(dots) |> dots_to_string()

  @doc """
  Make one fold along a horizontal (:y) or vertical (:x) line.

  ## Examples:

                           x: 3         y: 2      x: 3, y: 2
           ..O|.O          .OO|        ..O|.O        .OO|
           .O.|.O          .O.|        OO.|.O        OO.|
           ---+--          ---+        ---+--        ---+
           O..|.O          OO.|

      iex> dots = MapSet.new([{2, 0}, {5, 0}, {1, 1}, {5, 1}, {0, 3}, {5, 3}])
      iex> fold_one({:x, 3}, dots)
      MapSet.new([{1, 0}, {2, 0}, {1, 1}, {0, 3}, {1, 3}])
      iex> fold_one({:y, 2}, dots)
      MapSet.new([{2, 0}, {5, 0}, {0, 1}, {1, 1}, {5, 1}])
  """
  def fold_one({:x, line}, dots), do: fold_one(dots, fn {x, y} -> {line - abs(line - x), y} end)
  def fold_one({:y, line}, dots), do: fold_one(dots, fn {x, y} -> {x, line - abs(line - y)} end)
  def fold_one(dots, fold), do: dots |> Enum.map(&fold.(&1)) |> Enum.into(MapSet.new())

  @doc """
  Make several folds along horizontal and vertical lines.

  ## Example:

      iex> dots = MapSet.new([{2, 0}, {5, 0}, {1, 1}, {5, 1}, {0, 3}, {5, 3}])
      iex> fold_all([x: 3, y: 2], dots)
      MapSet.new([{1, 0}, {2, 0}, {0, 1}, {1, 1}])
  """
  def fold_all(folds, dots), do: Enum.reduce(folds, dots, &fold_one/2)

  @doc """
  Convert a set of dots to a printable string.

  ## Example:

      iex> dots_to_string(MapSet.new([{1, 0}, {0, 1}, {2, 1}, {1, 2}]))
      " █ \\n█ █\\n █ "
  """
  def dots_to_string(dots) do
    {width, _} = Enum.max_by(dots, &elem(&1, 0))
    {_, height} = Enum.max_by(dots, &elem(&1, 1))

    for y <- 0..height do
      for x <- 0..width do
        if {x, y} in dots, do: "█", else: " "
      end
    end
    |> Enum.map_join("\n", &Enum.join/1)
  end

  def main(args) do
    Enum.map(args, fn path -> AOC.solve(path, &parse/1, &part1/1, &part2/1) end)
  end
end
